src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -0,0 +1,460 @@
+        // If the value is busy then return busy
+        if (expressedState == ExpressedStateEnum::kSmokeAlarm || expressedState == ExpressedStateEnum::kCOAlarm ||
+            expressedState == ExpressedStateEnum::kTesting || expressedState == ExpressedStateEnum::kInterconnectSmoke ||
+            expressedState == ExpressedStateEnum::kInterconnectCO)
+        {
+            status = Status::Busy;
+        }
+        else
+        {
+            success = opHandler(endpointId);
+            if (!success)
+            {
+                status = Status::Failure;
+            }
+        }
+    }
+    else
+    {
+        status = Status::Failure;
+    }
+
+    commandObj->AddStatus(commandPath, status);
+    return true;
+}
+
+template <typename T>
+void SmokeCoAlarmServer::SendEvent(EndpointId endpointId, T & event)
+{
+    EventNumber eventNumber;
+    auto err = LogEvent(event, endpointId, eventNumber);
+
+    if (CHIP_NO_ERROR != err)
+    {
+        ChipLogError(Zcl, "Failed to log event: err=%" CHIP_ERROR_FORMAT ", event_id=" ChipLogFormatMEI, err.Format(),
+                     ChipLogValueMEI(event.GetEventId()));
+    }
+}
+
+template <typename T>
+bool SmokeCoAlarmServer::GetAttribute(EndpointId endpointId, AttributeId attributeId,
+                                      EmberAfStatus (*getFn)(EndpointId endpointId, T * value), T & value) const
+{
+    EmberAfStatus status = getFn(endpointId, &value);
+    bool success         = (EMBER_ZCL_STATUS_SUCCESS == status);
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to read SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x",
+                     ChipLogValueMEI(attributeId), to_underlying(status));
+    }
+    return success;
+}
+
+template <typename T>
+bool SmokeCoAlarmServer::SetAttribute(EndpointId endpointId, AttributeId attributeId,
+                                      EmberAfStatus (*setFn)(EndpointId endpointId, T value), T value)
+{
+    EmberAfStatus status = setFn(endpointId, value);
+    bool success         = (EMBER_ZCL_STATUS_SUCCESS == status);
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to write SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x",
+                     ChipLogValueMEI(attributeId), to_underlying(status));
+    }
+    return success;
+}
+
+// =============================================================================
+// Cluster commands callbacks
+// =============================================================================
+
+bool emberAfSmokeCoAlarmClusterSelfTestRequestCallback(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                       const Commands::SelfTestRequest::DecodableType & commandData)
+{
+    return SmokeCoAlarmServer::Instance().HandleRemoteSelfTestRequest(commandObj, commandPath,
+                                                                      emberAfPluginSmokeCoAlarmSelfTestRequestCommand);
```

Comment:
If we are hardcoding this anyway, why not have HandleRemoteSelfTestRequest call this function directly instead of passing the pointer, etc?  Either way, though.